File: app/src/main/java/com/example/phinh/grocery/Shopping/adapters/AdapterPromotionShop.java
package com.example.phinh.grocery.Shopping.adapters;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;


import com.example.phinh.grocery.R;
import com.example.phinh.grocery.Shopping.activities.AddPromotionCodeActivity;
import com.example.phinh.grocery.Shopping.models.ModelPromotion;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.Objects;

public class AdapterPromotionShop extends RecyclerView.Adapter<AdapterPromotionShop.HolderPromotionShop> {

    private final Context context;
    private final ArrayList<ModelPromotion> promotionArrayList;

    private final ProgressDialog progressDialog;
    private final FirebaseAuth firebaseAuth;

    public AdapterPromotionShop(Context context, ArrayList<ModelPromotion> promotionArrayList) {
        this.context = context;
        this.promotionArrayList = promotionArrayList;

        firebaseAuth = FirebaseAuth.getInstance();

        progressDialog = new ProgressDialog(context);
        progressDialog.setTitle("Vui lòng đợi");
        progressDialog.setCanceledOnTouchOutside(false);
    }

    @NonNull
    @Override
    public HolderPromotionShop onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        //inflate layout row_promotion_shop.xml
        View view = LayoutInflater.from(context).inflate(R.layout.row_promotion_shop, parent, false);

        return new HolderPromotionShop(view);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull final HolderPromotionShop holder, int position) {
        //get data
        final ModelPromotion modelPromotion = promotionArrayList.get(position);
        String id = modelPromotion.getId();
        String timestamp = modelPromotion.getTimestamp();
        String description = modelPromotion.getDescription();
        String promoCode = modelPromotion.getPromoCode();
        String promoPrice = modelPromotion.getPromoPrice();
        String expireDate = modelPromotion.getExpireDate();
        String minimumOrderPrice = modelPromotion.getMinimumOrderPrice();

        //set data
        holder.descriptionTv.setText(description);
        holder.promoPriceTv.setText(promoPrice);
        holder.minimumOrderPriceTv.setText(minimumOrderPrice);
        holder.promoCodeTv.setText("Code: "+promoCode);
        holder.expireDateTv.setText("Expire Date: "+expireDate);

        /*handle click, show Edit/Delete dialog*/
        holder.itemView.setOnClickListener(view -> editDeleteDialog(modelPromotion, holder));
    }

    private void editDeleteDialog(final ModelPromotion modelPromotion, HolderPromotionShop holder) {
        //options to display in dialog
        String[] options = {"Chỉnh sửa", "Xóa"};
        //dialog
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle("Vui lòng chọn các tùy chọn")
                .setItems(options, (dialogInterface, i) -> {
                    //handle clicks
                    if (i==0){
                        //Edit clicked
                        editPromoCode(modelPromotion);
                    }
                    else  if (i==1){
                        //Delete clicked
                        deletePromoCode(modelPromotion);
                    }
                })
                .show();
    }

    private void deletePromoCode(ModelPromotion modelPromotion) {
        //show progress bar
        progressDialog.setMessage("Xóa mã khuyến mãi...");
        progressDialog.show();

        DatabaseReference ref = FirebaseDatabase.getInstance().getReference("Users");
        ref.child(Objects.requireNonNull(firebaseAuth.getUid())).child("Promotions").child(modelPromotion.getId())
                .removeValue()
                .addOnSuccessListener(aVoid -> {
                    //deleted
                    progressDialog.dismiss();
                    Toast.makeText(context, "Xóa thành công...", Toast.LENGTH_SHORT).show();
                })
                .addOnFailureListener(e -> {
                    //failed deleting
                    progressDialog.dismiss();
                    Toast.makeText(context, ""+e.getMessage(), Toast.LENGTH_SHORT).show();
                });
    }

    private void editPromoCode(ModelPromotion modelPromotion) {
        //start and pass data to AddPromotionCodeActivity to edit
        Intent intent = new Intent(context, AddPromotionCodeActivity.class);
        intent.putExtra("promoId", modelPromotion.getId()); //will use id to update promo code
        context.startActivity(intent);
    }

    @Override
    public int getItemCount() {
        return promotionArrayList.size();
    }

    //view holder class
    static class HolderPromotionShop extends RecyclerView.ViewHolder{

        private final TextView promoCodeTv, promoPriceTv, minimumOrderPriceTv, expireDateTv, descriptionTv;

        public HolderPromotionShop(@NonNull View itemView) {
            super(itemView);

            //init ui views
            //views of row_promotion_shop.xml
            ImageView iconIv = itemView.findViewById(R.id.iconIv);
            promoCodeTv = itemView.findViewById(R.id.promoCodeTv);
            promoPriceTv = itemView.findViewById(R.id.promoPriceTv);
            minimumOrderPriceTv = itemView.findViewById(R.id.minimumOrderPriceTv);
            expireDateTv = itemView.findViewById(R.id.expireDateTv);
            descriptionTv = itemView.findViewById(R.id.descriptionTv);
        }
    }
}

File: app/src/main/java/com/example/grocery/adapters/AdapterPromotionShop.kt
package com.example.grocery.adapters

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.example.grocery.activities.AddPromotionCodeActivity
import com.example.grocery.adapters.AdapterPromotionShop.HolderPromotionShop
import com.example.grocery.models.ModelPromotion
import com.example.grocery.R
import com.example.grocery.activities.CurrencyFormatter
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import java.util.Objects

class AdapterPromotionShop(
    private val context: Context,
    private val promotionArrayList: ArrayList<ModelPromotion?>?
) : RecyclerView.Adapter<HolderPromotionShop>() {
    private val progressDialog: ProgressDialog
    private val firebaseAuth: FirebaseAuth

    init {
        firebaseAuth = FirebaseAuth.getInstance()
        progressDialog = ProgressDialog(context)
        progressDialog.setTitle("Vui lòng đợi")
        progressDialog.setCanceledOnTouchOutside(false)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HolderPromotionShop {
        //inflate layout row_promotion_shop.xml
        val view = LayoutInflater.from(context).inflate(R.layout.row_promotion_shop, parent, false)
        return HolderPromotionShop(view)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: HolderPromotionShop, position: Int) {
        //get data
        val modelPromotion = promotionArrayList!![position]
        val id = modelPromotion?.id
        val timestamp = modelPromotion?.timestamp
        val description = modelPromotion?.description
        val promoCode = modelPromotion?.promoCode
        val promoPrice = modelPromotion?.promoPrice
        val expireDate = modelPromotion?.expireDate
        val minimumOrderPrice = modelPromotion?.minimumOrderPrice

// Assuming promoPrice and minimumOrderPrice are Strings and might contain "$"
// If they are already Double, you can simplify the parsing part.

        val promoPriceDouble = promoPrice?.replace("$", "")?.toDoubleOrNull() ?: 0.0
        holder.promoPriceTv.text = CurrencyFormatter.format(promoPriceDouble)

        val minimumOrderPriceDouble = minimumOrderPrice?.replace("$", "")?.toDoubleOrNull() ?: 0.0
        holder.minimumOrderPriceTv.text = CurrencyFormatter.format(minimumOrderPriceDouble)

        //set data
        holder.descriptionTv.text = description
        holder.promoCodeTv.text = "Mã số: $promoCode"
        holder.expireDateTv.text = "Ngày hết hạn: $expireDate"

        /*handle click, show Edit/Delete dialog*/holder.itemView.setOnClickListener { view: View? ->
            editDeleteDialog(
                modelPromotion!!,
                holder
            )
        }
    }

    private fun editDeleteDialog(modelPromotion: ModelPromotion, holder: HolderPromotionShop) {
        //options to display in dialog
        val options = arrayOf("Chỉnh sửa", "Xóa")
        //dialog
        val builder = AlertDialog.Builder(context)
        builder.setTitle("Vui lòng chọn các tùy chọn")
            .setItems(options) { dialogInterface: DialogInterface?, i: Int ->
                //handle clicks
                if (i == 0) {
                    //Edit clicked
                    editPromoCode(modelPromotion)
                } else if (i == 1) {
                    //Delete clicked
                    deletePromoCode(modelPromotion)
                }
            }
            .show()
    }

    private fun deletePromoCode(modelPromotion: ModelPromotion) {
        //show progress bar
        progressDialog.setMessage("Xóa mã khuyến mãi...")
        progressDialog.show()
        val ref = FirebaseDatabase.getInstance().getReference("Users")
        ref.child(Objects.requireNonNull(firebaseAuth.uid).toString()).child("Promotions").child(
            modelPromotion.id!!
        )
            .removeValue()
            .addOnSuccessListener { aVoid: Void? ->
                //deleted
                progressDialog.dismiss()
                Toast.makeText(context, "Xóa thành công...", Toast.LENGTH_SHORT).show()
            }
            .addOnFailureListener { e: Exception ->
                //failed deleting
                progressDialog.dismiss()
                Toast.makeText(context, "" + e.message, Toast.LENGTH_SHORT).show()
            }
    }

    private fun editPromoCode(modelPromotion: ModelPromotion) {
        //start and pass data to AddPromotionCodeActivity to edit
        val intent = Intent(context, AddPromotionCodeActivity::class.java)
        intent.putExtra("promoId", modelPromotion.id) //will use id to update promo code
        context.startActivity(intent)
    }

    override fun getItemCount(): Int {
        return promotionArrayList!!.size
    }

    //view holder class
    class HolderPromotionShop(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val promoCodeTv: TextView
        val promoPriceTv: TextView
        val minimumOrderPriceTv: TextView
        val expireDateTv: TextView
        val descriptionTv: TextView

        init {

            //init ui views
            //views of row_promotion_shop.xml
            val iconIv = itemView.findViewById<ImageView>(R.id.iconIv)
            promoCodeTv = itemView.findViewById(R.id.promoCodeTv)
            promoPriceTv = itemView.findViewById(R.id.promoPriceTv)
            minimumOrderPriceTv = itemView.findViewById(R.id.minimumOrderPriceTv)
            expireDateTv = itemView.findViewById(R.id.expireDateTv)
            descriptionTv = itemView.findViewById(R.id.descriptionTv)
        }
    }
}
